### PR TITLE
Prototype: Binary Distribution for sLSTM extension

### PR DESCRIPTION
**Pull request by daidahao**
This PR proposes a prototype for building and distributing the CUDA-accelerated sLSTM extension as a binary wheel. The main goal is to remove CUDA source compilation from the common default-sLSTM path while keeping the existing JIT compilation path for custom configurations.

The prototype:
- Builds the default sLSTM CUDA kernel into the `xlstm._slstm` extension during wheel creation.
- Uses the precompiled extension when the runtime `sLSTMCellConfig` matches the default configuration used by the sLSTM-only [TiRex model](https://github.com/NX-AI/tirex).
- Falls back to the existing JIT compilation path when the precompiled extension is unavailable or the configuration does not match the precompiled macro set.
- Documents the remaining trade-offs around platform support, CUDA/PyTorch compatibility, customization, and test coverage.

## Feedback Requested
- Is this fallback behavior acceptable for a first binary-distribution prototype?
- Which distribution strategy should we pursue for CUDA-specific wheels: one PyPI wheel, CUDA-specific package names, GitHub releases, or a custom package index?
- Should we keep the current pybind11 extension for the prototype, or prioritize moving toward a PyTorch dispatcher registration (`TORCH_LIBRARY`) to enable CPython stable ABI.

cc: @lukfischer @martinloretzzz @kpoeppel @maximilianmbeck

## What Changed
- Added a `scikit-build-core` build backend in `pyproject.toml` for building the C++/CUDA extension into the wheel.
- Added a top-level `CMakeLists.txt` that defines the `_slstm` extension target, source files, CUDA architectures, PyTorch/CUDA/pybind11 linkage, and default sLSTM compile-time macros.
- Updated `xlstm/blocks/slstm/cell.py` so the CUDA backend first tries to import `xlstm._slstm` for the supported default configuration and otherwise uses the existing JIT compilation flow.

## Prototype Build Matrix
The package still supports Python 3.10+ and PyTorch 2.0+, but this prototype wheel itself is only built and tested for the following matrix:

| Component | Prototype target |
| --- | --- |
| OS/architecture | Linux x86_64 |
| Build image | Ubuntu 24.04.4 LTS, glibc 2.39 |
| CUDA toolkit | CUDA 12.6 |
| CUDA architectures | Current CMake default: 7.5, 8.0, 8.6, 9.0+PTX |
| Python | Built and tested with CPython 3.13 |
| PyTorch | Built and tested with PyTorch 2.11.0+cu126 |

Compatibility with other Python minors, PyTorch versions, CUDA versions, operating systems, glibc versions, and architectures still needs dedicated wheel builds and tests.

## Runtime Behavior
When `backend="cuda"` is selected, `sLSTMCellCUDA.instance()` checks whether the `sLSTMCellConfig` matches the default macro set compiled into the wheel:

- `num_heads == 4`
- `num_states == 4`
- `dtype_b == dtype_a == "float32"`
- `dtype == dtype_r == dtype_w == dtype_g == dtype_s == "bfloat16"`
- `num_gates == 4`
- `gradient_recurrent_clipval is None`

If all conditions match, xlstm attempts to import `xlstm._slstm`. If the import succeeds, no source compilation is needed at runtime. If the import fails, or if the configuration differs, xlstm falls back to the existing JIT compilation path.

## Build Process
The build process uses `scikit-build-core` to compile the C++/CUDA extension and package it as a wheel. The key steps include:
1. Set up `pyproject.toml` with the necessary build dependencies and configuration.
2. Add `CMakeLists.txt` to define the build process for the extension, which includes:
    - Set up pybind11 for C++ bindings.
    - Set up CUDA toolkit and language support.
    - Add source files (`.cu`, `.cc`) and link PyTorch CMake, CUDA, and pybind11 libraries.
    - **IMPORTANT**: Define sLSTM-specific macros (e.g., `SLSTM_NUM_HEADS`) to match the default configuration (`xlstm/blocks/slstm/cell.py`'s `sLSTMCellConfig` defaults).
    - Declare compile options to match the JIT-compiled version (`xlstm/blocks/slstm/src/cuda_init.py` extra compile flags).
3. Update `cell.py` to use the precompiled extension when the configuration matches the default, e.g., TiRex.
4. Use `uv` as the build frontend to create the wheel in `dist/`.

Here is a minimal shell script to build the wheel on Ubuntu 24.04 with CUDA 12.6:

```bash
# Install CUDA toolkit (skip if already installed)
wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2404/x86_64/cuda-keyring_1.1-1_all.deb
sudo dpkg -i cuda-keyring_1.1-1_all.deb
sudo apt-get update
sudo apt-get install -y cuda-toolkit-12-6

# Install uv (skip if already installed)
curl -LsSf https://astral.sh/uv/install.sh | sh

# Set up Python environment
uv python pin 3.13
uv sync --cache-dir /tmp/uv_cache

# Build the wheel
uv build --wheel --cache-dir /tmp/uv_cache
```

## Validation
To verify the functionality of the binary extension, I adapted [Quick Start from TiRex](https://github.com/NX-AI/tirex#quick-start) to use the precompiled extension in a TiRex context. TiRex uses the default sLSTM configuration, so it runs the precompiled path.

```python
import numpy as np
import torch
from tirex import load_model, ForecastModel
import time

N = 10

def main():
    model: ForecastModel = load_model(
        "NX-AI/TiRex",
        backend="cuda",
    )
    data = torch.linspace(0, 4 * torch.pi, 192).unsqueeze(0).repeat(5, 1)  # Sample Data (5 time series with length 192)
    data = torch.sin(data)  # Apply sine function to create a time series pattern
    context = data[:, :128]  # Use the first 128 points as context for forecasting

    def forecast():
        return model.forecast(context=context, prediction_length=64)

    # Warm-up
    forecast()

    # Measure time taken for N forecasts
    start_time = time.perf_counter()
    for _ in range(N):
        forecast_result = forecast()
    end_time = time.perf_counter()
    
    print(f"Forecast Result Shape: {forecast_result[0].shape}")
    avg_time = (end_time - start_time) / N
    print(f"Average time taken for forecasting: {avg_time:.4f} seconds")

    print(f"Forecast result: {forecast_result[0][0][:10][4].cpu().numpy()}")

if __name__ == "__main__":
    main()
```

### Test Environment
- Ubuntu 24.04.4 LTS
- CUDA 12.6
- Python 3.13.9
- PyTorch 2.11.0+cu126
- xlstm 2.0.5 (precompiled with the sLSTM extension)
- tirex-ts 1.4.1
- GPU: NVIDIA GeForce RTX 4090

### Speed

The timing below excludes model loading and includes one warm-up forecast before measurement.

| Backend | Average forecast time | Relative speed |
| --- | ---: | ---: |
| Precompiled CUDA extension | 0.0170 seconds | 12.2x faster than PyTorch |
| PyTorch backend, no JIT compilation | 0.2071 seconds | baseline |

Precompiled CUDA output:
```
Forecast Result Shape: torch.Size([5, 64, 9])
Forecast result: [0.6563942  0.6613323  0.66419464 0.66798973 0.6715489  0.67404324
 0.6760385  0.6804709  0.68497825]
```

PyTorch backend output:
```
Forecast Result Shape: torch.Size([5, 64, 9])
Forecast result: [0.65570503 0.66066885 0.66354257 0.6673481  0.67091465 0.67341584
 0.67542136 0.6798661  0.68439025]
```

Different backends produce similar forecast results, but the precompiled CUDA backend is significantly faster than the vanilla PyTorch implementation.

### Fidelity

To verify the correctness of the precompiled extension against the JIT-compiled CUDA version, I ran the same test on the same platform (same OS, GPU) but in a `conda` environment (to avoid system-level CUDA conflicts):

- CUDA 12.6
- Python 3.13.13
- PyTorch 2.11.0+cu126
- xlstm 2.0.5 (PyPI version)
- tirex-ts 1.4.1 (PyPI version)

```
Forecast Result Shape: torch.Size([5, 64, 9])
Average time taken for forecasting: 0.0177 seconds
Forecast result: [0.6563942  0.6613323  0.66419464 0.66798973 0.6715489  0.67404324
 0.6760385  0.6804709  0.68497825]
```

The results are consistent with the precompiled extension, confirming that the binary distribution produces correct forecasts while preventing compilation from source.

This is still an informal fidelity check. A production-ready validation suite should compare the precompiled CUDA extension, JIT-compiled CUDA extension, and PyTorch backend with explicit tolerances such as maximum absolute error and mean absolute error.

## Current Limitations and Solutions
1. **Platform Support**: The current prototype only builds for Linux x86_64 with Glibc 2.39. This can solved by building wheels for major platforms and Glibc versions (Linux) via CI workflows like `cibuildwheel` and docker images. This includes:
    - Linux with Glibc 2.28+ (`manylinux_2_28` covers most modern Linux distributions) and `x86_64` and `aarch64` architectures.
    - Windows support can be added later, but it requires additional build configuration and testing.
    - Non-compiled binaries for other platforms (e.g., macOS) that lack CUDA support.

    A `cibuildwheel` CI workflow for building binary wheels for all major platforms can be found [here](https://github.com/daidahao/nanobind_example).
2. **Python Compatibility**: The prototype is built against Python 3.13, whereas other Python minors must be separately built against. Unlike nanobind, pybind11 does not support stable CPython ABI. We may use `cibuildwheel` to build wheels for all Python minors (3.10, 3.11, 3.12, 3.13, 3.14) to ensure broad compatibility. To use CPython ABI, we can ditch pybind11 and use the [`TORCH_LIBRARY` macro](https://docs.pytorch.org/tutorials/advanced/cpp_custom_ops.html) to create a PyTorch extension.
3. **CUDA Compatibility**: While CUDA provides backward compatibility, the prototype when built against CUDA 12.6 is dynamically linked to `cublas.so.12`. Just like before, we can use CIs to build separate wheels for different CUDA versions (12.6, 12.8, 13.0 currently supported by PyTorch). However, the distribution will become an issue as PyPI does not distinguish CUDA versions. We can either:
    - Distribute 13.0-only wheel on PyPI (same as PyTorch) but distribute others via GitHub releases or self hosted index (PyTorch does this for CUDA [12.6](https://download.pytorch.org/whl/cu126) and [12.8](https://download.pytorch.org/whl/cu128)).
    - Create separate packages for different CUDA versions (e.g., `xlstm-cu126`, `xlstm-cu128`, `xlstm-cu130`) and publish all on PyPI.
    - Build and bundle all three CUDA versions into a single wheel, and conditionally load the appropriate CUDA version at runtime. This is complicated to set up because CMake relies on system-level CUDA toolkit and testing all three versions is non-trivial.
4. **PyTorch Compatibility**: The prototype is built against PyTorch 2.11.0, but it may not be compatible with other PyTorch versions. Fortunately, PyTorch provides [stable ABI from 2.10 onwards](https://docs.pytorch.org/tutorials/advanced/cpp_custom_ops.html). We may adapt `xlstm/blocks/slstm/src/cuda/slstm.cc` to use the new C++ API to build a single wheel compatible with 2.10+.
5. **Customized Configurations**: The binary distribution only supports the default sLSTM configuration (e.g., 4 heads, 4 gates as macros) with the precompiled extension. Other configurations will still trigger JIT compilation. We can convert some of those macros (e.g., `SLSTM_NUM_HEADS`) into runtime parameters and use a single precompiled extension to support multiple configurations. However, other macros like `SLSTM_DTYPE_B` still require separate builds.
6. **Testing**: The prototype currently only includes a basic test on a single platform as I don't have multiple GPUs. We can set up CIs to run tests across different platforms and CUDA versions for each built wheel, ensuring robustness and correctness.
7. **Distribution**: The prototype currently does not include a distribution strategy, but we can use PyPI as the main channel and GitHub releases as complementary as suggested above.

### CODE REVIEW

**Comment by daidahao** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by martinloretzzz** [comment]
Thanks a lot for the contribution. I'm quite busy this week, I'll have a look on it next week.

**Comment by martinloretzzz** [comment]
Thanks a lot for all the great work. Just to be sure, if I'm on a platform where the precompiled version isn't supported even when the shapes match to use the precompiled, this will result in the import error and then the standard version is used?

This PR looks good, but in the future I'd like to use the huggingface kernels to prebuild for all configurations and backends.
